[CGP-439] [AST] Added a separate subproject, strings and logging

effectfully opened the pull request:
Please forgive me this.

effectfully commented:
Okay, ready for review. 

There are many changes here, but it's mostly administrative bookkeeping and refactoring. The most important bits are in these two consecutive notes: https://github.com/input-output-hk/plutus/blob/ede001a0b2a1db773847551034e78ff9bc6faa8f/language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs#L228

We need it merged as soon as possible, so suggestions like "rename that" or "move this to that file" I'll probably take as future tasks (but feel free to point out anything that does not suit your taste).

This is still not something final, but we're close now.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs:
```diff
@@ -72,20 +72,22 @@ kindOfTypeBuiltin TySize       = sizeToType
     Left  (_::RenameError ()) -> throwing _TypeError $ InternalTypeErrorE ann $ OpenTypeOfBuiltin ty con
-    Right annTyOfName         -> pure $ NormalizedType annTyOfName
+    -- That's quite inefficient, but is there anything we can do about that?
+    Right annTyOfName         -> normalizeType annTyOfName
```
We want to prenormalize types of dynamic built-in names. Previously they were normalized, but now we have dynamic types which can be non-normalized. That's a future work.

michaelpj commented on language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs:
```diff
@@ -207,6 +181,130 @@ newtype DynamicBuiltinNameMeanings = DynamicBuiltinNameMeanings
+        evaluateCek anEnvironment (foldList {dyn} {unit} (\(r : unit) -> emit) unitval list)
+
+After evaluation finishes, we read a Haskell list from the 'IORef'
+(which requires another 'unsafePerformIO') and return it.
```
What would happen if we made the monad that the CEK machine runs in a monad transformer, and allowed builtin evaluation to have effects in that monad?

Then, for example, `emit` could use a `STRef` instead of an `IORef` (or indeed we could keep using an `IORef` and run over IO, but that seems overkill). This would also be a way of allowing other kinds of effects inside builtins.

This is a very sketchy suggestion so far, it's possible that the types make this difficult. I think it would be okay if we instantiated our monad differently in different cases: so for the "main" CEK evaluation it's `Identity`, but for these `evaluateCek` calls you use to convert the builtins to Haskell values you could use `ST`.

effectfully replied:
Interesting thoughts. I'll probably play with this stuff later. But we need some motivation for it. `emit` is a hack that is supposed to vanish at some point. Do we need effects for something else?

michaelpj replied:
Well, I'm pretty uncomfortable with the `unsafePerformIO` stuff in here, so I'd feel a lot more like I understood what was going on and that I wasn't going to accidentally break things by subtly changing how GHC complies things.

In addition, I think we're planning to keep the trace effect in the mockchain for a while, so that could be modelled as happening in an appropriate base monad

effectfully replied:
> Well, I'm pretty uncomfortable with the `unsafePerformIO` stuff in here

So am I, but I have only two days left and the main stuff is not done yet. I'll happily try out your suggestion later.

michaelpj replied:
Well, okay. Maybe I'll try it while you're away.

effectfully replied:
That's not fair!

michaelpj replied:
I can't let you have all the fun :p

michaelpj commented on language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs:
```diff
@@ -207,6 +181,130 @@ newtype DynamicBuiltinNameMeanings = DynamicBuiltinNameMeanings
+we want to resume evaluation during computation of a dynamic built-in application, we just call the
+received evaluator
+
+(3) seems best, so it's what is implemented.
```
Personally I think (1) has the advantage of being conceptually simplest, but I can see implementing it might be a pain.

Should (3) also be the basis of an abstract interface to the evaluators in general?

michaelpj commented on plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs:
```diff
@@ -65,6 +68,13 @@ data Frame
+runCekM :: CekEnv -> CekM a -> Either CekMachineException a
+runCekM = flip runReaderT
+
+-- TODO: use lenses ffs.
```
perhaps you should do this :)

effectfully replied:
Done.

michaelpj commented on plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs:
```diff
@@ -167,32 +177,46 @@ applyEvaluate funVarEnv _         con fun                    arg =
                     ConstAppError   err ->
                         throwError $ MachineException (ConstAppMachineError err) term
 
+-- TODO: we must catch exceptions here. Or maybe reflect them in the `KnownDynamicBuiltinType` class.
```
Wouldn't this be a lot simpler if `Evaluate` reflected the non-throwing interface to the machines? Then you would just get back an `Either` and it would be straightforward.

effectfully replied:
Yes, but I haven't tried it yet. Besides, I again think evaluators should throw and that making the CEK machine non-throwing was a mistake. It really should be the case that the CEK machine receives a term wrapped in `newtype Checked = Checked (Term TyName Name a)` and any error then in an internal one and we get rid of those `Either`s, checks and other garbage.

michaelpj replied:
Well, except it's hard to compose things that throw, as we're discovering here. I still think they shouldn't throw, but :shrug:

effectfully replied:
Here we're discovering that it's hard to make throwing and non-throwing stuff play well together, no?

effectfully replied:
Fixed. Now the entire built-in applications machinery runs monadically. That was surprising easy to do. And one can instantiate the monad to `Identity` (for the CK machine) or to `Either CekMachineException` (for the CEK machine). This is also a step towards the `CekT` transformer.

michaelpj replied:
Right - my point is that it's easy to compose things that don't throw, and it's easy to make things that don't throw throw at the edges of the program. But it's annoying to compose a thing that does throw and a think that doesn't.

So I claim that if we used the non-throwing version then our composition would be easier.

michaelpj replied:
Nice.

effectfully commented:
Once thing I forgot to emphasize:

> We do not type check terms here, because type checking of nested dynamic built-in types simply
> does not work. The type checker can't be quickly repaired, so we keep it like this for now.

The reason for that is probably that the annotation machinery assigns kinds to types beforehand, but we generate some types on the fly and they can contain free variables. I do not know how true this is, but I'm getting errors from the annotater. This needs to be handled carefully later.

effectfully commented:
I added a complete example.
See [this](https://github.com/input-output-hk/plutus/blob/32d25091a09f741af771916c21d0b33c21138f13/language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/OnChain.hs) file for handlers of dynamic built-ins (I added some docs there) and [this](https://github.com/input-output-hk/plutus/blob/58a2afbb720a97070528d9ced2ca3494aa6c3f6c/plutus-core-interpreter/test/DynamicBuiltins/Logging.hs) file for examples. Here is one of them:

```haskell
test_logInt :: TestTree
test_logInt = testCase "logInt" $ do
    let size = 4
        term
            = Apply () dynamicLog
            . Apply () (TyInst () dynamicIntToString (TyInt () size))
            $ Constant () (BuiltinInt () size 1)

    let eval1 = evaluateHandlersCek (handleDynamicIntToString . handleDynamicLog)
    let eval2 = evaluateHandlersCek (handleDynamicLog . handleDynamicIntToString)
    (logs1, _) <- eval1 $ OnChain term
    (logs2, _) <- eval2 $ OnChain term
    logs1 @?= ["1"]
    logs2 @?= ["1"]
```

In short, we define a PLC function `dynamicIntToString` by wrapping Haskell's `show` as a dynamic built-in name and then we can call `dynamicLog (dynamicIntToString x)` as many times as we want (this expression returns `unitval` as the result). That's easy.

The complicated part is how we evaluate such terms. Basically, you have to provide a handler for each of the names. Functions that do not do anything effectful can be lifted to handlers easily via `handleDynamicByMeaning`,e.g.

```haskell
handleDynamicIntToString :: OnChainHandler "intToString" f r r
handleDynamicIntToString = handleDynamicByMeaning dynamicIntToStringMeaning
```

Convoluted stuff like logging-via-unsafePerformIO-and-IORefs is more complicated, but the good thing is that pure and impure handlers compose with each other in any order using `.` (I do not recommend at the moment composing several impure handlers, because god knows how that plays with `unsafePerformIO`)

In the end you pass a pipe of such handlers to `evaluateHandlersCek` and it performs evaluation.

michaelpj commented:
Wow, you weren't joking when you said I'd be horrified by the way you're handling names.

I'm not sure how easy this is going to be to fit into the serialization framework still. We're just serializing `Program TyName Name ()` at the moment and reading it back out - I don't know how we could persist the appropriate `Symbol` information that we'd need to make everything line up....

effectfully commented:
> We're just serializing Program TyName Name () at the moment and reading it back out - I don't know how we could persist the appropriate Symbol information that we'd need to make everything line up....

As Note [Names at the type-level] explains, you have to know the allowed names beforehand on the chain. I.e. you do not read `Program TyName Name ()` back, you read `OnChain allowedNames TyName Name ()` back.

effectfully commented:
@michaelpj, I'd like to merge this. The only new thing since your last approval is that hairy handling of names (I believe it makes sense), but it's not used anywhere except tests. Do you have objections?

We still can continue the discussion about how this is supposed to be used on the blockchain.

michaelpj commented:
No, I think this is in a reasonable state now where we can either go forward with using it or pause it to come back to later. Merge when you're ready (maybe squash merge).

effectfully commented on language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Emit.hs:
```diff
@@ -0,0 +1,52 @@
+
+-- We do not type check terms here, because type checking of nested dynamic built-in types simply
+-- does not work. The type checker can't be quickly repaired, so we keep it like this for now.
+withEmitEvaluateBy
```
This is kinda obsolete now that we have `OnChain.hs`, but it's *much* clearer than what is used there and it has `evaluate` which may (or may not) be crucial due to this whole `unsafePerformIO` thing, so I'll keep it here for now.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs:
```diff
@@ -207,6 +181,132 @@ newtype DynamicBuiltinNameMeanings = DynamicBuiltinNameMeanings
 
+type Evaluator f m = DynamicBuiltinNameMeanings -> f TyName Name () -> m EvaluationResult
+
+type Evaluate m = ReaderT (Evaluator Term m) m
```
It's monadic right now only because the CEK machine is monadic, but I believe the latter should throw exceptions. However if we manage to build the `CekT` transformer, having monadic evaluators will be useful of course. In general, the current design is fine, but we may want to revisit it in future.
